### DISCUSSION TITLE
How do I add code generation support for another language, or modify exiting language?

### DISCUSSION THREAD

**Discussion by Joebeazelman** [StateSmith/StateSmith#374] [Q&A] [answered]
I'm really digging StateSmith! I would like to support another programming language, but I couldn't find where the state machine code is generated from in the repositories. 

Where do I find it? Also, does it use a templating facility like T4, or mustache?

**Comment by adamfk** [comment, marked as answer]
Awesome!

The current code generation has two stages.
* we first generate Generic Intermediate Language (GIL) code according to our selected **Algorithm**
* we then transpile the GIL to the target language

The **algorithm** is all about how the state machine code works. Does it use function pointers? Switch statements? A table? See more here: https://github.com/StateSmith/StateSmith/wiki/Algorithms

The GIL is just a restricted subset of C# so that we can use Roslyn tools to aid in transpiling.

I will help get you started. This will be our first community added language! 🎉 


<br>

# Step 0
Open a new issue with the language to support.


<br>

# Step 1 - manual code translation
Let's start simple. We'll manually create the desired output for a tiny hierarchical state machine.

* Take the below state machine and generate code using a currently supported language. Maybe pick one closest to your target language. Let's assume it is JavaScript.
* Manually convert the generated JavaScript into the new language.
* and we can continue from there.

![image](https://github.com/user-attachments/assets/30706831-8f80-4896-ae75-1afae723f898)

```plantuml
@startuml LightSm

state OFF
state ON_GROUP {
    state ON1
    state ON2
}

' ////////// STATE HANDLERS //////////
[*] -> OFF

OFF: enter / off();
OFF --> ON1: INC

ON_GROUP --> OFF: OFF

ON1: enter / blueLaser();
ON1 --> ON2: INC
ON1 --> OFF: DIM

ON2: enter / yellowLaser();
ON2: enter / count = 0;
ON2: 1. INC / count++;
ON2 --> ON1: DIM

@enduml
```

I'm excited to see us support a new language :)

Thanks!
